src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -231,39 +233,33 @@ private static Binder GetEnclosingBinder(SyntaxNode node, int position, Binder r
                         binder = rootBinder.GetBinder(current);
                     }
                 }
-                else if (current.Kind() == SyntaxKind.EqualsValueClause)
+                else if (kind == SyntaxKind.EqualsValueClause)
                 {
                     binder = rootBinder.GetBinder(current);
                 }
-                else if (current.Kind() == SyntaxKind.Attribute)
+                else if (kind == SyntaxKind.Attribute)
                 {
                     binder = rootBinder.GetBinder(current);
                 }
-                else if (current.Kind() == SyntaxKind.ArrowExpressionClause)
+                else if (kind == SyntaxKind.ArrowExpressionClause)
                 {
                     binder = rootBinder.GetBinder(current);
                 }
-                else if (current is ExpressionSyntax && 
+                else if (current is ExpressionSyntax &&
                             ((current.Parent as LambdaExpressionSyntax)?.Body == current ||
                              (current.Parent as SwitchStatementSyntax)?.Expression == current ||
                              (current.Parent as CommonForEachStatementSyntax)?.Expression == current))
                 {
                     binder = rootBinder.GetBinder(current);
                 }
-                else if (current is VariableComponentSyntax &&
-                             (current.Parent as ForEachComponentStatementSyntax)?.VariableComponent == current)
+                else if ((kind == SyntaxKind.DeclarationExpression || kind == SyntaxKind.TupleExpression) &&
+                             (current.Parent as ForEachVariableStatementSyntax)?.Variable == current)
```

Comment:
>(current.Parent as ForEachVariableStatementSyntax)?.Variable == current [](start = 29, length = 71)

Is this special case really needed? It feels like we would get to the parent anyway.